lib/vbt/provider.ex

```diff
@@ -0,0 +1,74 @@
+defmodule VBT.Provider do
+  alias Ecto.Changeset
+
+  @type adapter :: module
+  @type params :: %{param_name => param_spec}
+  @type param_name :: atom
+  @type param_spec :: %{type: type, default: value}
+  @type type :: :string | :integer | :float | :boolean
+  @type value :: String.t() | number | boolean | nil
+  @type data :: %{param_name => value}
+
+  # ------------------------------------------------------------------------
+  # API
+  # ------------------------------------------------------------------------
+
+  @spec fetch_all(adapter, params) :: {:ok, data} | {:error, [String.t()]}
+  def fetch_all(adapter, params) do
+    types = Enum.into(params, %{}, fn {name, spec} -> {name, spec.type} end)
+
+    data =
+      params
+      |> Stream.zip(adapter.values(Map.keys(types)))
+      |> Enum.into(%{}, fn {{param, opts}, provided_value} ->
+        value = if is_nil(provided_value), do: opts.default, else: provided_value
+        {param, value}
+      end)
+
+    {%{}, types}
+    |> Changeset.cast(data, Map.keys(types))
+    |> Changeset.validate_required(Map.keys(types), message: "is missing")
+    |> case do
+      %Changeset{valid?: true} = changeset -> {:ok, Changeset.apply_changes(changeset)}
+      %Changeset{valid?: false} = changeset -> {:error, changeset_error(adapter, changeset)}
+    end
```

Comment:
Cool example of `Ecto.Changeset` use for general parameter validation - just a heads-up if people miss it.

Reply:
Yes! It's called [Schemalless Changeset](https://hexdocs.pm/ecto/Ecto.Changeset.html#module-schemaless-changesets), and it can be a useful technique in e.g. phoenix controllers, when we want to normalize parameters before sending them to context functions. It's not needed in resolvers, b/c there the heavy lifting is done by abinthe.